src/bin/cut.rs

```diff
@@ -0,0 +1,525 @@
+            }
+        }
+        return None;
+    }
+}
+
+/// Set a boolean array element to true, extending the array if necessary.
+fn toggle(v: &mut Vec<bool>, index: usize) {
+    if index + 1 > v.len() {
+        v.resize(index + 1, false);
+    }
+    v[index] = true;
+}
+
+/// Fill a range (including first and last element) in a boolean array to true,
+/// extending the array if necessary.
+fn fill(v: &mut Vec<bool>, begin: usize, end: usize) {
+    // This could be more eficient by breaking it up into
+    // differente cases and filling with true and false
+    // but I do not think is worth it.
+    if end > v.len() {
+        v.resize(end, false);
+    }
+    for index in begin..end {
+        v[index] = true;
+    }
+}
+
+/// Write selected bytes from a reader.
+fn cut_bytes<R: Read, W: Write>(input: R, output: W, selection: &Selection) {
+    let reader = io::BufReader::new(input);
+    let mut writer = io::BufWriter::new(output);
+
+    for line in reader.lines() {
+        let line = line.unwrap();
+        let bytes = line.bytes();
+        for el in selection.select_from(bytes) {
+            let _ = writer.write(&[el]);
+        }
+        let _ = writer.write(b"\n");
+    }
+}
+
+/// Write selected characters from a reader.
+fn cut_characters<R: Read, W: Write>(input: R, output: W, selection: &Selection) {
+    let reader = io::BufReader::new(input);
+    let mut writer = io::BufWriter::new(output);
+
+    for line in reader.lines() {
+        let line = line.unwrap();
+        let chars = line.chars();
+        for el in selection.select_from(chars) {
+            let _ = write!(&mut writer, "{}", el);
+        }
+        let _ = writer.write(b"\n");
+    }
+}
+
+/// Write selected fields from a reader.
+///
+/// The delimiter is also printed when there are more than 1 fields.
+///
+/// # arguments
+///
+/// * `delimiter` - Indicates how the fields are delimited
+/// * `skip_if_missing` - If true, lines not containing the field delimiter will be skipped.
+fn cut_fields<R: Read, W: Write>(input: R, output: W, selection: &Selection,
+                                 delimiter: &str, skip_if_missing: bool) {
+    let reader = io::BufReader::new(input);
+    let mut writer = io::BufWriter::new(output);
+
+    for line in reader.lines() {
+        let line = line.unwrap();
+        if !line.contains(delimiter) {
+            if !skip_if_missing {
+                let _ = writeln!(&mut writer, "{}", line);
+            }
+            continue;
+        }
+        let fields = line.split(delimiter);
+
+        // The first element is separated as a way to include the delimiter
+        // in the output if there are more than two elements.
+        let mut it = selection.select_from(fields);
+        if let Some(el) = it.next() {
+            let _ = write!(&mut writer, "{}", el);
+        } else {
+            continue;
+        }
+        for el in it {
+            let _ = write!(&mut writer, "{}{}", delimiter, el);
```

Comment:
Good question! The reason is that using formatters (e.g. `write!` or `println!`) will allocate memory for building the string.